src/log_surgeon/finite_automata/DfaTransition.hpp

```diff
@@ -0,0 +1,70 @@
+#ifndef LOG_SURGEON_FINITE_AUTOMATA_DFATRANSITION_HPP
+#define LOG_SURGEON_FINITE_AUTOMATA_DFATRANSITION_HPP
+
+#include <cstdint>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+
+#include <log_surgeon/finite_automata/RegisterOperation.hpp>
+#include <log_surgeon/finite_automata/StateType.hpp>
+
+namespace log_surgeon::finite_automata {
+template <StateType state_type>
+class DfaState;
+
+template <StateType state_type>
+class DfaTransition {
+public:
+    DfaTransition() = default;
+
+    DfaTransition(std::vector<RegisterOperation> reg_ops, DfaState<state_type>* dest_state)
+            : m_reg_ops{std::move(reg_ops)},
+              m_dest_state{dest_state} {}
+
+    [[nodiscard]] auto get_reg_ops() const -> std::vector<RegisterOperation> { return m_reg_ops; }
```

Comment:
```suggestion
    [[nodiscard]] auto get_reg_ops() const -> std::vector<RegisterOperation> const& { return m_reg_ops; }
```
How about returning a const reference?